server/src/test/java/com/lunchtime/service/impl/OrderServiceImplTest.java

```diff
@@ -0,0 +1,217 @@
+
+    @Test
+    public void personCanNotMakeOrderWhenStartTimeBiggerThanFinishTime() {
+        LocalDateTime start = LocalDateTime.now().plusDays(3);
+        LocalDateTime finish = LocalDateTime.now().plusMinutes(40).plusDays(2);
+        Date startDate = Date.from(start.atZone(ZoneId.systemDefault()).toInstant());
+        Date finishDate = Date.from(finish.atZone(ZoneId.systemDefault()).toInstant());
+
+        Order order = Order.builder()
+            .table(restaurantTable)
+            .person(person)
+            .startTime(startDate)
+            .finishTime(finishDate)
+            .visitors(5)
+            .build();
+
+        when(orderRepository.findAllOrdersByTableInTime(order.getTable().getId(), startDate, finishDate))
+            .thenReturn(Collections.emptyList());
+        when(orderRepository.save(order)).thenReturn(order);
+
+        Order createdOrder = orderServiceImpl.saveOrder(order);
+
+        assertNull(createdOrder);
+    }
+
+    @Test
+    public void personCanNotMakeOrderWhenStartTimeLessThanCurrentTime() {
+        LocalDateTime start = LocalDateTime.now().minusDays(1);
+        LocalDateTime finish = LocalDateTime.now().plusMinutes(40).plusDays(2);
+        Date startDate = Date.from(start.atZone(ZoneId.systemDefault()).toInstant());
+        Date finishDate = Date.from(finish.atZone(ZoneId.systemDefault()).toInstant());
+
+        Order order = Order.builder()
+            .table(restaurantTable)
+            .person(person)
+            .startTime(startDate)
+            .finishTime(finishDate)
+            .visitors(5)
+            .build();
+
+        when(orderRepository.findAllOrdersByTableInTime(order.getTable().getId(), startDate, finishDate))
+            .thenReturn(Collections.emptyList());
+        when(orderRepository.save(order)).thenReturn(order);
+
+        Order createdOrder = orderServiceImpl.saveOrder(order);
+
+        assertNull(createdOrder);
+    }
+
+    @Test
+    public void personCanNotMakeOrderWhenPersonDoesNotExist() {
+        LocalDateTime start = LocalDateTime.now().minusDays(2);
+        LocalDateTime finish = LocalDateTime.now().plusMinutes(40).plusDays(2);
+        Date startDate = Date.from(start.atZone(ZoneId.systemDefault()).toInstant());
+        Date finishDate = Date.from(finish.atZone(ZoneId.systemDefault()).toInstant());
+
+        Person anotherPerson = new Person();
+        anotherPerson.setId(2L);
+
+        Order order = Order.builder()
+            .table(restaurantTable)
+            .person(anotherPerson)
+            .startTime(startDate)
+            .finishTime(finishDate)
+            .visitors(5)
+            .build();
+
+        when(orderRepository.findAllOrdersByTableInTime(order.getTable().getId(), startDate, finishDate))
+            .thenReturn(Collections.emptyList());
+        when(orderRepository.save(order)).thenReturn(order);
+
+        Order createdOrder = orderServiceImpl.saveOrder(order);
+
+        assertNull(createdOrder);
+    }
+}
```

Comment:
I'm not close to testing. What I see logically, test method name should be easier and clear for its reader. Use comments to exaplain what test's purpose is.

Then, it seems that Date can be created easily and moved to `setup()`. If we'll need another date in our test, we can just modify it whereas, as well as `order`. 

`when` statements are repeated multiply. I guess there is way to improve it. you have.

I will discover Testing more soon and give you further feedback. Also I'm interested in team's developers opinions about this testing.

Reply:
> Use comments to exaplain what test's purpose is.

Absolutely no! Don't use comments in code, your code must be selfdocumented. And in test cases, method names are documentation of project

Reply:
`personCanMakeOrderInSpecificTimeIfRestaurantTableIsAvailable` <- this is not the best way to explain what test method does. Also it is recommended to use only one `Mock` in testing class

Reply:
Why? What do you don't understand?

> Also it is recommended to use only one Mock in testing class
I'm newby in testing. Could you show examples of right solutions?

Reply:
My business logic depends on all those Mocks, how can I should use all necessary repositories without Mock?

Reply:
> Then, it seems that Date can be created easily and moved to setup(). If we'll need another date in our test, we can just modify it whereas, as well as order.

I've tried to do it, but it doesn't work

Reply:
I'm also new in testing, so I can be mistaken. **[This](https://stackoverflow.com/a/17810004/13334444)** tells us that:
_According to the principle of Test only one thing per test, there may be several stubs in one test, but generally there is only one mock._

> I've tried to do it, but it doesn't work

**[Try](https://examples.javacodegeeks.com/core-java/junit/junit-setup-teardown-example/)** `tearDown()` to do a prepration and post clean up process for each of the JUnit method test call.

Reply:
Thanks! Very useful annotation `@After`, I couldn't find it earlier
But, I can't change all Mocks to Stubs

Reply:
I'm also learning testing at the moment. Minute ago I found **[this](https://stackoverflow.com/questions/3459287/whats-the-difference-between-a-mock-stub)**, it may help us.